### ISSUE TITLE
`std_instead_of_alloc` misses items from the implicit prelude

### ISSUE THREAD

**Issue by bushrat011899** [rust-lang/rust-clippy#17304]
### Summary

Items brought in from the implicit prelude are not checked by the `std_instead_of_core` family of lints. For items in `core::prelude` this isn't really an issue, since marking a crate as `#![no_std]` will switch to the `core` implicit prelude, so no errors will occur.

However, for items provided by `alloc` that are only available in the `std` implicit prelude, this is quite annoying. For the following items, their usage will not be flagged by `std_instead_of_alloc`, so they will not be caught until they become a compilation error when switching to `no_std`:

* `Box`
* `String`
* `ToOwned`
* `ToString`
* `Vec`
* `format`
* `vec`

### Lint Name

std_instead_of_core

### Reproducer

I tried this code:

```rust
#![warn(clippy::std_instead_of_alloc)]
#![allow(unused_imports)]

use Box;
use String;
use ToOwned;
use ToString;
use Vec;
use format;
use vec;
```

I expected to see this happen:

Seven instances of `std_instead_of_alloc`, one per `use` statement. These should say to switch from `Box` to `alloc::boxed::Box`, etc.

Instead, this happened:

No lints emitted.

### Version

```text
rustc 1.98.0-nightly (485ec3fbc 2026-06-10)
binary: rustc
commit-hash: 485ec3fbcc12fa14ef6596dabb125ad710499c9e
commit-date: 2026-06-10
host: x86_64-pc-windows-msvc
release: 1.98.0-nightly
LLVM version: 22.1.6
```

**Comment by Halloloid** [comment]
I've reproduced the issue and found that `use Vec`; has `path.res = alloc::vec::Vec`, but `first_segment.res = Err`, so the current `check_path` exits early. I'm looking how implicit prelude imports should be detected. Is there an existing rustc API I should be using?

**Comment by bushrat011899** [comment]
> I've reproduced the issue and found that `use Vec`; has `path.res = alloc::vec::Vec`, but `first_segment.res = Err`, so the current `check_path` exits early. I'm looking how implicit prelude imports should be detected. Is there an existing rustc API I should be using?

Sadly I'm not aware of any way to detect where an identity comes from in the late pass. A test case you probably want to compare against is this:

```rust
fn a() {
    let _ = Vec::<()>::new();
}

fn b() {
    extern crate alloc;
    use alloc::vec::Vec;
    let _ = Vec::<()>::new();
}
```

`a` should trigger `std_instead_of_alloc`, but `b` should not. Aside from updating the compiler to track implicit prelude information and expose it on the `TyCtxt`, the only way I can think to do this is to just recreate the identity resolution logic.

**Comment by Halloloid** [comment]
Thanks! That makes sense. I'll investigate whether there's any existing compiler information I can use before trying to recreate the resolution logic. If I don't find one, would you prefer a rustc-side change to expose implicit prelude information, or would a Clippy-side solution be acceptable?

**Comment by bushrat011899** [comment]
> Thanks! That makes sense. I'll investigate whether there's any existing compiler information I can use before trying to recreate the resolution logic. If I don't find one, would you prefer a rustc-side change to expose implicit prelude information, or would a Clippy-side solution be acceptable?

I feel like a nice solution would be something like `TyCtxt::module_children` which also works for blocks, not just module-likes. That way you could check iterate the ancestors of a particular `HirId` and see if any have a `use` item that brings it into scope. If not, it must come from the implicit prelude.

You can do this right now for modules, but it doesn't consider block-like scopes, so any `use` statements in a function or local scope are invisible and would generate a false positive.

**Comment by Halloloid** [comment]
I've spent some time investigating this issue and was able to reproduce it. I found that implicit prelude imports like `use Vec`; resolve to `alloc::vec::Vec`, but the corresponding Path has only a single segment whose `res` is `Err`, so the current check_path logic exits before reaching the existing lint implementation.

I also explored the surrounding HIR and discussed possible approaches, but I'm currently blocked on the compiler internals needed to reliably distinguish implicit prelude imports from local use statements.

Since I don't think I can make further progress right now, I'm releasing my assignment so someone with more rustc internals experience can continue. Thanks for the guidance! @bushrat011899
